enhancements/authentication/pod-security-admission-autolabeling.md

```diff
@@ -0,0 +1,293 @@
+  - to be able to properly assess SA capabilities in controlled namespaces
+- **SCCs**
+  - to properly assess restrictive PSa levels per affected namespaces
+
+The generic control loop of the Controller performs the following actions:
+1. list all controlled (opted-in by a label) namespaces
+2. for each namespace:
+    1. get all instances of known workload types
+    1. associate SCCs to the workloads based on the SAs running them, create a
+       SCC <-> []SA associations
+    1. order observed SCCs most-privileged first (according to PSa privilege levels
+       they map to)
+    1. evaluate the namespace's PSa privilege level based on the most privileged
+       SCC that is allowed to be used by its SA
+    1. modify the namespace's PSa labels based on the enforce level observed in the
+       previous step
+
+There is another control loop reacting only to changes of namespaces which performs
+all the steps in 2.
+
+#### Label Synchronization Opting-in
+
+The Controller described in [PSa Label Synchronization Controller](#psa-label-synchronization-controller)
+works on an opt-in basis, meaning that users need to specifically sign themselves up
+if they wanted to stick with just SCCs.
+
+The opt-in is driven by the `security.openshift.io/scc.podSecurityLabelSync: “true”`
+label, namespaces containing it will have their PSa labels reconciled by
+the Controller.
+
+The opt-in nature of the Controller would cause current workloads to break.
+The choice of opting-in is therefore only left upon the users that are
+capable of creating the `Namespace` objects directly. Such a user can clearly
+set the appropriate labels themselves.
+
+The rest of the users rely on `Namespace` creation delegated via the OpenShift
+`Project` API. The default `Project` template should therefore be modified so
+that it by default includes the `security.openshift.io/scc.podSecurityLabelSync: “true”`
+label. This ensures that the decision to use PSa as the main authorization
+mechanism is left to privileged users only and the workloads of the rest
+of the users keeps working.
+
+Simply updating the `Project` template is not enough to keep older workloads
+working during upgrades. For that purpose, a controller should exist in an
+4.(n-1) version that annotates all namespaces created via the `Project` API.
+The controller assumes such a namespace creation by the "openshift.io/requester"
+annotation on the namespace, this annotation is added by the default `Project`
+template. The 4.n version is the version when the PSa gets turned on by default
+in restricted level, 4.(n-1) is the version that precedes it.
+
+#### SCC to PSa Level Transformation
+
+The automatic PSa label synchronization requires us to map SCCs directly to a PSa
+level.
+
+In order to be able to map SCCs to PSa level automatically, it is necessary to
+introspect each field of such an SCC.
+
+SCC fields each tell the SCC authorizer how to either:
```

Comment:
Have you written this mapping down somewhere?

Reply:
I've got a small PoC which shows some SCC shortcomings that need to be addressed (see FIXMEs in that project): https://github.com/stlaz/sccpsaconversion

Reply:
I think describing this mapping here is worthwhile.

Reply:
I'm going to open a follow-up PR for that